### PR TITLE
compiler/natives/src/strconv: Use js string conversion utilities to improve performance

### PR DESCRIPTION
**Pull request by lologarithm**
Added native implemntations for Atoi and Itoa. This generates smaller and faster code.

Before (using go stdlib)
```
$ gopherjs test strconv --run=Atoi --bench=Atoi
goos: linux
goarch: js
BenchmarkAtoi/Pos/7bit    	 2000000	       606 ns/op
BenchmarkAtoi/Pos/26bit   	 2000000	       628 ns/op
BenchmarkAtoi/Pos/31bit   	  200000	      8700 ns/op
BenchmarkAtoi/Neg/7bit    	 2000000	       660 ns/op
BenchmarkAtoi/Neg/26bit   	 2000000	       735 ns/op
BenchmarkAtoi/Neg/31bit   	  200000	      9030 ns/op
PASS
ok  	strconv	12.110s
```

After (using new js native)
```
$ gopherjs test strconv --run=Atoi --bench=Atoi
goos: linux
goarch: js
BenchmarkAtoi/Pos/7bit    	10000000	       111 ns/op
BenchmarkAtoi/Pos/26bit   	10000000	       150 ns/op
BenchmarkAtoi/Pos/31bit   	10000000	       214 ns/op
BenchmarkAtoi/Neg/7bit    	10000000	       138 ns/op
BenchmarkAtoi/Neg/26bit   	10000000	       158 ns/op
BenchmarkAtoi/Neg/31bit   	10000000	       222 ns/op
PASS
ok  	strconv	11.481s
```

I also wrote a small Itoa benchmark since there wasn't one in std lib. This uses 123456789 as the test number
```
gopherjs test --bench=.
goos: linux
goarch: js
BenchmarkGoItoa      1000000          1217 ns/op
BenchmarkJSItoa     10000000           142 ns/op
PASS
```

### CODE REVIEW

File: compiler/natives/src/strconv/atoi.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		return 0, syntaxError(fnAtoi, s)
+	}
+	jsValue := js.Global.Call("Number", s, 10)
+	if !js.Global.Call("isFinite", jsValue).Bool() {
+		return 0, syntaxError(fnAtoi, s)
+	}
+	// Bounds checking
+	floatval := jsValue.Float()
+	if floatval > maxInt32 {
+		return 1<<31 - 1, rangeError(fnAtoi, s)
```
Comment (flimzy) [inline compiler/natives/src/strconv/atoi.go]: Does it make sense to replace `1<<31 -1` with the `maxInt32` const here?  And similar for `minInt32` below?

File: compiler/natives/src/strconv/atoi.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+const maxInt32 float64 = 1<<31 - 1
+const minInt32 float64 = -1 << 31
+
+// Atoi returns the result of ParseInt(s, 10, 0) converted to type int.
+func Atoi(s string) (int, error) {
+	const fnAtoi = "Atoi"
+	if len(s) == 0 {
+		return 0, syntaxError(fnAtoi, s)
+	}
+	jsValue := js.Global.Call("Number", s, 10)
```
Comment (flimzy) [inline compiler/natives/src/strconv/atoi.go]: I believe JS's `Number` constructor is far more forgiving than Go's `Atoi` implementation. Is it therefore reasonable to validate that `s` contains only Go-acceptable numbers?

Some examples of values that `Number()` accepts but `Atoi()` should not:

- "0x10"
- "10.2"
- "0b11"
- "0o10"

Comment (lologarithm) [reply compiler/natives/src/strconv/atoi.go]: Validate doesn't seem to have hurt perf at all -- pushing up the changes.

Comment (flimzy) [reply compiler/natives/src/strconv/atoi.go]: Looks like a good improvement. And glad to hear it doesn't have a noticeable impact on performance!

**Comment by lologarithm** [comment]
@shurcooL another optimization PR for you :)

**Comment by flimzy** [comment]
Superceded by #1101

**Pull request closed without merging**